mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java

```diff
@@ -310,6 +322,60 @@ public void setFlagsShouldNotChangeFlagsOfAnotherMessageInSameMailbox() throws E
     }
 
     @Test
+    public void setFlagsShouldNotChangeFlagsWhenEmptyMailboxes() throws Exception {
+        Flags newFlags = new Flags(Flags.Flag.SEEN);
+        MessageId messageId1 = testingData.persist(mailbox1.getMailboxId(), FLAGS);
+
+        messageIdManager.setFlags(newFlags, MessageManager.FlagsUpdateMode.ADD, messageId1, ImmutableList.<MailboxId>of(), session);
+
+        List<Flags> flags = FluentIterable
+            .from(messageIdManager.getMessages(ImmutableList.of(messageId1), FetchGroupImpl.MINIMAL, session))
+            .transform(getFlags())
+            .toList();
+
+        assertThat(flags).hasSize(1);
+        assertThat(flags.get(0)).isEqualTo(FLAGS);
+    }
+
+    @Test
+    public void setFlagsShouldNotChangeFlagsWhenMessageDoesNotBelongToTheMailboxes() throws Exception {
+        Flags newFlags = new Flags(Flags.Flag.SEEN);
+        MessageId messageId1 = testingData.persist(mailbox1.getMailboxId(), FLAGS);
+
+        messageIdManager.setFlags(newFlags, MessageManager.FlagsUpdateMode.ADD, messageId1, ImmutableList.of(mailbox2.getMailboxId()), session);
+
+        List<Flags> flags = FluentIterable
+            .from(messageIdManager.getMessages(ImmutableList.of(messageId1), FetchGroupImpl.MINIMAL, session))
+            .transform(getFlags())
+            .toList();
+
+        assertThat(flags).hasSize(1);
+        assertThat(flags.get(0)).isEqualTo(FLAGS);
+    }
+
+    @Test
+    public void setFlagsShouldChangeFlagsWhenMessageBelongsToTheMailboxes() throws Exception {
+        Flags newFlags = new Flags(Flags.Flag.SEEN);
+        MessageId messageId1 = testingData.persist(mailbox1.getMailboxId(), FLAGS);
+        MessageId messageId2 = testingData.persist(mailbox2.getMailboxId(), FLAGS);
+
+        messageIdManager.setFlags(newFlags, MessageManager.FlagsUpdateMode.ADD, messageId1, ImmutableList.of(mailbox1.getMailboxId(), mailbox2.getMailboxId()), session);
+
+        final List<Map.Entry<MessageId, Flags>> entries = FluentIterable.from(messageIdManager.getMessages(ImmutableList.of(messageId1, messageId2), FetchGroupImpl.MINIMAL, session))
+            .transform(toMapEntryOfFlags())
+            .toList();
+        ImmutableMap.Builder<MessageId, Flags> builder = ImmutableMap.builder();
+        for (Map.Entry<MessageId, Flags> entry : entries) {
+            builder.put(entry);
+        }
+        Map<MessageId, Flags> flags = builder.build();
```

Comment:
I tried with toMap but it will build the Map with key is MessageResult (Object in list) but here we need MessageId (Property of object in list)